clients/adapter-kafka/src/main/java/org/eclipse/hono/adapter/client/command/kafka/KafkaBasedInternalCommandConsumer.java

```diff
@@ -260,9 +274,22 @@ void handleCommandMessage(final KafkaConsumerRecord<String, Buffer> record) {
         final CommandContext commandContext = new KafkaBasedCommandContext(command, currentSpan);
 
         if (commandHandler != null) {
-            LOG.trace("using [{}] for received command [{}]", commandHandler, command);
-            // command.isValid() check not done here - it is to be done in the command handler
-            commandHandler.handleCommand(commandContext);
+            // partition index and offset here are related to the *tenant-based* topic the command was originally received in
+            // therefore they are stored with the tenant as key
+            final Map<Integer, Long> lastHandledPartitionOffsets = lastHandledPartitionOffsetsPerTenant
+                    .computeIfAbsent(command.getTenant(), k -> new HashMap<>());
+            final Long lastHandledOffset = lastHandledPartitionOffsets.get(commandPartition);
+            if (lastHandledOffset != null && commandOffset <= lastHandledOffset) {
```

Comment:
The commands are forwarded by the command routers to this topic. The command records are handled by a void method `mapAndDelegateIncomingCommandMessage`. Inside this method we are making several future calls (get tenant, get adapter instance and then send command). It seems to me that we cannot ensure that the commands are forwarded in the same order we have received. In such cases, there is a chance of losing commands when we compare the offsets using `commandOffset <= lastHandledOffset`.

Reply:
> It seems to me that we cannot ensure that the commands are forwarded in the same order we have received. 

If I understand correctly, this can even mess up the order of commands to a particular device. That would not be acceptable. So, maybe the code in the command router needs to be changed. Then the logic here would work. The alternative, keeping _all_ handled offsets instead of only the last one,  seems not very practicable.

Reply:
> If I understand correctly, this can even mess up the order of commands to a particular device. That would not be acceptable. 

Well, strictly speaking, I think we don't guarantee that the input/output order is kept the same - also not in the case of AMQP commands and also not for telemetry/event messages as they are processed in the protocol adapters. The chances of this happening are however rather slim, I think (I find it difficult imagining a theoretical scenario where this would happen necessarily). But I agree that with the change in this PR, a message would be dropped in that case, and this should be prevented. I'll have a look at guaranteeing the order in the CommandRouter.

Reply:
> also not in the case of AMQP commands and also not for telemetry/event messages as they are processed in the protocol adapters

https://www.eclipse.org/hono/docs/api/telemetry-kafka/#consume-telemetry-data says: "Hono delivers messages containing telemetry data reported by a particular device in the same order that they have been received in". But maybe in the case of a device sending multiple messages at the same time, the execution in the adapter could be out of order.

Reply:
> But maybe in the case of a device sending multiple messages at the same time, the execution in the adapter could be out of order.

FMPOV this could only happen if a device has *multiple* connections to one or more protocol adapters at the same time. Within each Verticle instance, the messages would still be processed in the order that they have been received. This is due to the fact that each Verticle runs on a single Context.

Reply:
Yes, but what about external service invocations done during the processing of e.g. a telemetry message? I'm thinking about device registry and prometheus requests for example. Theoretically, one such request could take longer than the other. Although with caching in place and with the way such requests are handled in the corresponding components, thinking about a case where this would necessarily appear doesn't seem easy.

Reply:
Concerning possible scenarios relevant to this PR, there might be several commands from different devices, being received by the command router in quick succession and getting mapped to the same gateway and target protocol adapter instance. The number of data grid requests needed to do the mapping could be different for these devices, depending on the number of `via` entries of the device and whether the "last known gateway" is set (see the `CacheBasedDeviceConnectionInfo` class). Hence a later command could overtake an earlier command during processing in the command router. This would most probably involve commands for different devices, but if they are mapped to the same gateway and target adapter instance (and if they originate from the same topic partition), this becomes relevant with regard to the changes in this PR here.
To resolve this, I've now created #2617, ensuring that the command router sends the commands originating from the same topic partition in the order they were received.